### PR TITLE
bt-15: Create button shared component

### LINKED ISSUE #15: Create Button shared component
We need to add shared components which will be reused in the app (Button for this ticket )
Design link: [https://www.figma.com/file/Yd7USgP7PLvA0W794jHXWu/BSA_recruting?type=design&node-id=0-1&mode=design&t=Q2BWnU8ikNDNZ3Oh-0](url)

Image

### PR DESCRIPTION
**Pull request by mskmee**
image
image
image
image

Add fontawesome icons for future icons, set label like optional props for the same reason.
We can rewrite all styles from other components if we need,

### CODE REVIEW

File: mobile/package.json [hunk 1]
```diff
@@ -14,6 +14,10 @@
         "start:dev": "react-native start"
     },
     "dependencies": {
+        "@fortawesome/fontawesome-svg-core": "6.4.2",
+        "@fortawesome/free-brands-svg-icons": "6.4.2",
+        "@fortawesome/free-solid-svg-icons": "6.4.2",
+        "@fortawesome/react-native-fontawesome": "0.3.0",
```
Comment (katerynakostikova) [inline mobile/package.json]: Let's use [https://github.com/oblador/react-native-vector-icons](https://github.com/oblador/react-native-vector-icons) instead

Comment (mskmee) [reply mobile/package.json]: I have checked fort awesome icons, free icons didn't contain all icons that we needed.  As for  react-native-vector-icons
this library seems like very large, maybe it's not that we need. 
Maybe it will be better to add svg icon instead?.. Waiting for others to discuss

Comment (nikita-remeslov) [reply mobile/package.json]: In the web app, we use material icons. That is why I think - https://github.com/oblador/react-native-vector-icons will be good

File: mobile/src/bundles/common/components/button/button.tsx [hunk 2]
```diff
@@ -1,19 +1,122 @@
...
+
+const Button: React.FC<Properties> = ({
+    label = '',
+    labelStyles,
+    style,
+    isPrimary = true,
+    disabled = false,
+    icon,
+    iconProps,
+    ...properties
```
Comment (nikita-remeslov) [inline mobile/src/bundles/common/components/button/button.tsx]: Let's keep it simple :) 
```suggestion
    ...props
```

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]

File: mobile/src/bundles/common/components/button/button.tsx (hunk 2 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/button.tsx]: I wish, but linter said: "The variable `props` should be named `properties_`. A more descriptive name will do too.eslint[unicorn/prevent-abbreviations](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/v48.0.1/docs/rules/prevent-abbreviations.md)"

File: mobile/src/bundles/common/components/button/button.tsx [hunk 3]
```diff
@@ -1,19 +1,122 @@
...
+}) => {
+    const [isPressed, setIsPressed] = useState(false);
```
Comment (katerynakostikova) [inline mobile/src/bundles/common/components/button/button.tsx]: https://medium.com/timeless/a-simple-touch-feedback-with-the-new-pressable-component-b4d91c35456

File: mobile/src/bundles/common/components/button/button.tsx [hunk 4]
```diff
@@ -1,19 +1,122 @@
...
+    iconProps?: Omit<Props, 'icon'>;
+} & PressableProps;
+
+type StyleRecord = Record<string, unknown>;
+
+const Button: React.FC<Properties> = ({
+    label = '',
+    labelStyles,
+    style,
+    isPrimary = true,
```
Comment (katerynakostikova) [inline mobile/src/bundles/common/components/button/button.tsx]: Let's create an enum for button state, e.x:
```
const ButtonStatus = {
    FILLED: 'Filled',
    OUTLINE: 'Outline',
    GHOST: 'Ghost',
} as const;
```
Then in the switch case, you return button styles depending on the button type, the default value can be ButtonStatus.FILLED

File: mobile/src/bundles/common/components/button/button.tsx [hunk 5]
```diff
@@ -1,19 +1,122 @@
...
 
 import { Text } from '~/bundles/common/components/components';
 
 import { styles } from './styles';
 
 type Properties = {
-    label: string;
-    onPress: () => void;
-};
+    label?: string;
```
Comment (katerynakostikova) [inline mobile/src/bundles/common/components/button/button.tsx]:
```suggestion
    text: string;
```

File: mobile/src/bundles/common/components/button/button.tsx [hunk 6]
```diff
@@ -1,19 +1,122 @@
...
+    isPrimary?: boolean;
+    labelStyles?: StyleProp<ViewStyle>;
```
Comment (katerynakostikova) [inline mobile/src/bundles/common/components/button/button.tsx]:
```suggestion
```
You can change text styles depending on the button type

File: mobile/src/bundles/common/components/button/styles.ts [hunk 7]
```diff
@@ -1,13 +1,89 @@
 import { StyleSheet } from 'react-native';
 
 const styles = StyleSheet.create({
```
Comment (katerynakostikova) [inline mobile/src/bundles/common/components/button/styles.ts:5]: There are a lot of repeatings in styles, think of optimising them after adding button type and refactoring the pressed state

**Review by katerynakostikova — CHANGES_REQUESTED** [review]

File: mobile/src/bundles/common/components/button/button.tsx (hunk 3 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/button.tsx]: I've use this from the beginig, but as for me its too hard for understanding, when checks makes in jsx.

File: mobile/src/bundles/common/components/button/button.tsx (hunk 6 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/button.tsx]: it's already got right styles, relaid on props, this labelStyles is rewrite default values if we need.

File: mobile/src/bundles/common/components/button/button.tsx (hunk 5 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/button.tsx]: we go label property for button in fronedt, quess it should stay label.

File: mobile/src/bundles/common/components/button/button.tsx (hunk 3 above)
Comment (katerynakostikova) [reply mobile/src/bundles/common/components/button/button.tsx]: There's no need in creating a separate useState for pressed when it's already returned from styles, you can pass this value to the function for example

File: mobile/src/bundles/common/components/button/button.tsx (hunk 6 above)
Comment (katerynakostikova) [reply mobile/src/bundles/common/components/button/button.tsx]: As I see in UI, for now, there'll be no need in rewriting styles for text, but we can add it later if it'll be a need.

File: mobile/src/bundles/common/components/button/button.tsx (hunk 5 above)
Comment (katerynakostikova) [reply mobile/src/bundles/common/components/button/button.tsx]: Ok, but label should be required anyway

File: mobile/src/bundles/common/components/button/button.tsx (hunk 3 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/button.tsx]: noted, will fix

File: mobile/src/bundles/common/components/button/button.tsx (hunk 2 above)
Comment (katerynakostikova) [reply mobile/src/bundles/common/components/button/button.tsx]: Merge latest development and error will disappear

File: mobile/src/bundles/common/components/button/button.tsx (hunk 3 above)
Comment (katerynakostikova) [reply mobile/src/bundles/common/components/button/button.tsx]: Also, sometimes it's ok to add checks in jsx, e.x.:
`style={[styles.button, isError && styles.errorButton]}`
Maybe this option will be more readable for you

Comment (mskmee) [reply mobile/src/bundles/common/components/button/button.tsx]: fix

File: mobile/src/bundles/common/components/button/button.tsx (hunk 4 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/button.tsx]: fix

File: mobile/src/bundles/common/components/button/button.tsx (hunk 6 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/button.tsx]: fix

File: mobile/src/bundles/common/components/button/styles.ts (hunk 7 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/styles.ts:5]: fix

File: mobile/src/bundles/common/components/button/button.tsx (hunk 5 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/button.tsx]: fix

File: mobile/package.json (hunk 1 above)
Comment (mskmee) [reply mobile/package.json]: fix

File: mobile/src/bundles/common/components/button/button.tsx [hunk 8]
```diff
@@ -1,19 +1,95 @@
...
+    iconSize?: number;
+    buttonType?: ButtonTypeName;
+} & PressableProps;
+
+const iconDefaultSize = 32;
+
+const Button: React.FC<Properties> = ({
+    label,
+    style: pressableStyle,
+    iconName = 'add',
```
Comment (katerynakostikova) [inline mobile/src/bundles/common/components/button/button.tsx]:
```suggestion
    iconName ,
```
This field should not be required, you can add icon only if name is provided

File: mobile/src/bundles/common/components/button/button.tsx [hunk 9]
```diff
@@ -1,19 +1,95 @@
...
+    iconSize = iconDefaultSize,
+    buttonType = 'FILLED',
```
Comment (katerynakostikova) [inline mobile/src/bundles/common/components/button/button.tsx]:
```suggestion
    buttonType = ButtonType.FILLED
```

File: mobile/src/bundles/common/components/button/button.tsx [hunk 10]
```diff
@@ -1,19 +1,95 @@
...
+        FILLED: styles.button_filled_disabled,
+        OUTLINE: styles.button_outline_disabled,
+        GHOST: styles.content_disabled,
+    };
+
+    const pressedStyleLabel: Record<ButtonTypeName, StyleRecord> = {
+        FILLED: styles.label,
+        OUTLINE: styles.content_pressed,
+        GHOST: styles.content_pressed,
+    };
```
Comment (katerynakostikova) [inline mobile/src/bundles/common/components/button/button.tsx]: Let's create useMemo and return 1 object with all that styles

File: mobile/src/bundles/common/components/button/button.tsx [hunk 11]
```diff
@@ -1,19 +1,95 @@
...
-        <Pressable style={styles.btn} onPress={onPress}>
-            <Text style={styles.label}>{label}</Text>
+        <Pressable
+            disabled={disabled}
+            style={({ pressed }): StyleRecord[] => [
+                styles.button,
+                buttonStyleChoose[buttonType],
+                pressed ? buttonStylePressed[buttonType] : {},
+                disabled ? buttonStyleDisabled[buttonType] : {},
+                (pressableStyle as StyleRecord | undefined) ?? {},
```
Comment (katerynakostikova) [inline mobile/src/bundles/common/components/button/button.tsx]:
```suggestion
                pressed && buttonStylePressed[buttonType],
                disabled && buttonStyleDisabled[buttonType] ,
                pressableStyle,
```

File: mobile/src/bundles/common/components/button/styles.ts [hunk 12]
```diff
@@ -1,13 +1,63 @@
...
 
+import { Color } from '../../enums/enums';
+
 const styles = StyleSheet.create({
-    btn: {
-        padding: 10,
-        backgroundColor: 'blue',
+    button: {
+        paddingVertical: 15,
+        paddingHorizontal: 37,
```
Comment (katerynakostikova) [inline mobile/src/bundles/common/components/button/styles.ts]: Let's avoid magical numbers like 37, 43, 97 :)

File: mobile/src/bundles/common/components/button/styles.ts [hunk 13]
```diff
@@ -1,13 +1,63 @@
 import { StyleSheet } from 'react-native';
 
+import { Color } from '../../enums/enums';
+
 const styles = StyleSheet.create({
```
Comment (katerynakostikova) [inline mobile/src/bundles/common/components/button/styles.ts:5]: Remove some global styles and add it in component directly

File: mobile/src/bundles/common/components/button/button.tsx (hunk 10 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/button.tsx]: do you mean something like this: 
`   const componentStyles = useMemo(() => {
        return {
            [ButtonType.FILLED]: {
                style: styles.button_filled,
                pressed: styles.button_filled_pressed,
                disabled: styles.button_filled_disabled,
            },...
        };
    }, []);` 
   
    
image
 ?

Comment (katerynakostikova) [reply mobile/src/bundles/common/components/button/button.tsx]: Yes, something like this

File: mobile/src/bundles/common/components/button/button.tsx (hunk 8 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/button.tsx]: fix

File: mobile/src/bundles/common/components/button/button.tsx (hunk 9 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/button.tsx]: fix

File: mobile/src/bundles/common/components/button/button.tsx (hunk 11 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/button.tsx]: fix

File: mobile/src/bundles/common/components/button/button.tsx (hunk 10 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/button.tsx]: fix

File: mobile/src/bundles/common/components/button/styles.ts (hunk 12 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/styles.ts]: fix

File: mobile/src/bundles/common/components/button/styles.ts (hunk 13 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/styles.ts:5]: fix

File: mobile/src/bundles/auth/screens/auth.tsx [hunk 14]
```diff
@@ -57,6 +57,14 @@ const Auth: React.FC = () => {
...
         <>
             <Text>state: {dataStatus}</Text>
+            <Button buttonType={ButtonType.FILLED} label={'Filled'} />
+            <Button buttonType={ButtonType.OUTLINE} label={'Outlined'} />
+            <Button
+                buttonType={ButtonType.GHOST}
+                label={'GHOST'}
+                iconName={'add'}
+            />
+
```
Comment (katerynakostikova) [inline mobile/src/bundles/auth/screens/auth.tsx]:
```suggestion
            <Button buttonType={ButtonType.FILLED} label="Filled" />
            <Button buttonType={ButtonType.OUTLINE} label="Outlined"  />
            <Button
                buttonType={ButtonType.GHOST}
                label="GHOST" 
                iconName="add" 
            />

```

File: mobile/src/bundles/common/components/button/button.tsx [hunk 15]
```diff
@@ -1,19 +1,116 @@
...
+    iconName?: string;
+    iconSize?: number;
+    buttonType?: ButtonName;
+} & PressableProps;
+
+const iconDefaultSize = 32;
+
+const Button: React.FC<Properties> = ({
+    label,
+    style: pressAbleStyle,
```
Comment (katerynakostikova) [inline mobile/src/bundles/common/components/button/button.tsx]:
```suggestion
    style: pressableStyle,
```

File: mobile/src/bundles/common/components/button/button.tsx [hunk 16]
```diff
@@ -1,19 +1,116 @@
...
+    const isGhostButton = buttonType === ButtonType.GHOST;
     return (
-        <Pressable style={styles.btn} onPress={onPress}>
-            <Text style={styles.label}>{label}</Text>
+        <Pressable
+            disabled={disabled}
+            style={({ pressed }): StyleRecord[] => [
+                isGhostButton
+                    ? (globalStyles.borderRadius24,
+                      globalStyles.flexDirectionRow)
```
Comment (katerynakostikova) [inline mobile/src/bundles/common/components/button/button.tsx]: I guess we can add Icon not only to ghost type

File: mobile/src/bundles/common/components/button/button.tsx [hunk 17]
```diff
@@ -1,19 +1,116 @@
...
+            {({ pressed }): JSX.Element => (
+                <>
+                    {iconName && (
+                        <Icon
+                            name={iconName}
+                            size={iconSize}
+                            style={[
+                                styles.icon,
+                                pressed && styles.button_ghost_pressed,
+                                disabled && styles.content_disabled,
```
Comment (katerynakostikova) [inline mobile/src/bundles/common/components/button/button.tsx]: let's add some margin right to icon

File: mobile/src/bundles/common/styles/sizes.ts [hunk 18]
```diff
@@ -16,6 +16,12 @@ const SIZES = {
     borderRadius5: {
         borderRadius: 5,
     },
+    borderRadius24: {
+        borderRadius: 24,
+    },
+    borderWidth2: {
+        borderWidth: 2,
+    },
```
Comment (katerynakostikova) [inline mobile/src/bundles/common/styles/sizes.ts]: You can leave these styles in style.ts, since it's edge case for now related to only 1 button type

Comment (mskmee) [reply mobile/src/bundles/common/styles/sizes.ts]: should I create variables for this values or numbers will be ok?

File: mobile/src/bundles/auth/screens/auth.tsx (hunk 14 above)
Comment (mskmee) [reply mobile/src/bundles/auth/screens/auth.tsx]: fix

File: mobile/src/bundles/common/components/button/button.tsx (hunk 15 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/button.tsx]: fix

File: mobile/src/bundles/common/components/button/button.tsx (hunk 16 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/button.tsx]: fix

File: mobile/src/bundles/common/components/button/button.tsx (hunk 17 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/button.tsx]: fix

File: mobile/src/bundles/common/styles/sizes.ts (hunk 18 above)
Comment (mskmee) [reply mobile/src/bundles/common/styles/sizes.ts]: fix

File: mobile/src/bundles/common/components/button/button.tsx [hunk 19]
```diff
@@ -1,19 +1,115 @@
...
+import { ButtonType, TextCategory } from '../../enums/enums';
 import { styles } from './styles';
 
+type StyleRecord = Record<string, unknown> | undefined | null | false;
+
+type StylesProperties = {
+    style: Record<string, unknown>;
+    pressed: Record<string, unknown>;
+    disabled: Record<string, unknown>;
+    label: Record<string, unknown>;
```
Comment (katerynakostikova) [inline mobile/src/bundles/common/components/button/button.tsx]:
```suggestion
    basic: StyleProp<ViewStyle>;
    pressed: StyleProp<ViewStyle>;
    disabled: StyleProp<ViewStyle>;
    label: StyleProp<ViewStyle>;
```

File: mobile/src/bundles/common/components/button/button.tsx [hunk 20]
```diff
@@ -1,19 +1,115 @@
...
+import Icon from 'react-native-vector-icons/MaterialIcons';
 
-import { Text } from '~/bundles/common/components/components';
+import { Pressable, Text } from '~/bundles/common/components/components';
+import { globalStyles } from '~/bundles/common/styles/styles';
 
+import { ButtonType, TextCategory } from '../../enums/enums';
 import { styles } from './styles';
 
+type StyleRecord = Record<string, unknown> | undefined | null | false;
```
Comment (katerynakostikova) [inline mobile/src/bundles/common/components/button/button.tsx]: Can we simplify it?

File: mobile/src/bundles/common/components/button/button.tsx [hunk 21]
```diff
@@ -1,19 +1,115 @@
...
+                globalStyles.pv10,
+                globalStyles.ph25,
+                globalStyles.alignItemsCenter,
+                globalStyles.justifyContentCenter,
+
+                styles.button,
+                componentStyles[buttonType].style,
+                pressed && componentStyles[buttonType].pressed,
+                disabled && componentStyles[buttonType].disabled,
+                pressableStyle as Record<string, unknown>,
```
Comment (katerynakostikova) [inline mobile/src/bundles/common/components/button/button.tsx]:
```suggestion
                pressableStyle
```

File: mobile/src/bundles/common/components/button/styles.ts [hunk 22]
```diff
@@ -1,13 +1,57 @@
...
+const ghostIconRadius = 24;
+
 const styles = StyleSheet.create({
-    btn: {
-        padding: 10,
-        backgroundColor: 'blue',
-        alignItems: 'center',
+    button: {
+        borderColor: 'transparent',
+        borderWidth: borderWidth,
```
Comment (katerynakostikova) [inline mobile/src/bundles/common/components/button/styles.ts]:
```suggestion
        borderWidth: 2,
```

File: mobile/src/bundles/common/components/button/styles.ts [hunk 23]
```diff
@@ -1,13 +1,57 @@
...
+    button_filled_disabled: {
+        backgroundColor: Color.BACKGROUND,
+    },
+    button_outline_disabled: {
+        backgroundColor: '#FFFFFF',
+        borderColor: Color.INPUT,
+    },
+    button_ghost: {
+        backgroundColor: '#FFFFFF',
+        borderRadius: ghostIconRadius,
```
Comment (katerynakostikova) [inline mobile/src/bundles/common/components/button/styles.ts]:
```suggestion
        borderRadius: 24,
```

File: mobile/src/bundles/common/components/button/button.tsx (hunk 19 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/button.tsx]: fix

File: mobile/src/bundles/common/components/button/button.tsx (hunk 20 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/button.tsx]: you are absolutly right, fix it

File: mobile/src/bundles/common/components/button/button.tsx (hunk 21 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/button.tsx]: fix

File: mobile/src/bundles/common/components/button/styles.ts (hunk 22 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/styles.ts]: fix

File: mobile/src/bundles/common/components/button/styles.ts (hunk 23 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/styles.ts]: fix

File: mobile/src/bundles/common/components/button/button.tsx [hunk 24]
```diff
@@ -1,19 +1,116 @@
...
+
+const iconDefaultSize = 32;
 
-const Button: React.FC<Properties> = ({ label, onPress }) => {
+const Button: React.FC<Properties> = ({
+    label,
+    style: pressableStyle,
+    iconName,
+    iconSize = iconDefaultSize,
+    buttonType,
```
Comment (katerynakostikova) [inline mobile/src/bundles/common/components/button/button.tsx]:
```suggestion
    buttonType = ButtonType.Filled,
```

**Review by katerynakostikova — APPROVED** [review]

**Review by likeri29 — APPROVED** [review]

File: mobile/src/bundles/common/components/button/button.tsx (hunk 24 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/button.tsx]: If I've enter template condition i've got error next with conditions: 
image
`Unnecessary conditional, both sides of the expression are literal values.eslint[@typescript-eslint/no-unnecessary-condition](https://typescript-eslint.io/rules/no-unnecessary-condition)
(alias) const ButtonType: {
    readonly FILLED: "Filled";
    readonly OUTLINE: "Outline";
    readonly GHOST: "Ghost";
}
import ButtonType`
Guess linter was update. Maybe you can give addvice to me, how I can check button type in another way?

Comment (anogire) [reply mobile/src/bundles/common/components/button/button.tsx]: Maybe with ':'? 
buttonType: ButtonType.Filled,

Comment (mskmee) [reply mobile/src/bundles/common/components/button/button.tsx]: Unworthily ternary operator did`t help too :( Thanks for advice

Comment (anogire) [reply mobile/src/bundles/common/components/button/button.tsx]: One more chance ...
If it should just be a default value, then maybe buttonType = ButtonType.FILLED ?

Comment (mskmee) [reply mobile/src/bundles/common/components/button/button.tsx]: I've make type assertion and it's work, but looks ugly :( Thanks for advices ^) If anyone find better tell about it :)

File: mobile/src/bundles/common/components/button/button.tsx [hunk 25]
```diff
@@ -1,19 +1,116 @@
...
+                },
+                [ButtonType.GHOST]: {
+                    basic: styles.button_ghost,
+                    pressed: styles.button_ghost_pressed,
+                    disabled: styles.content_disabled,
+                    label: styles.content_pressed,
+                },
+            };
+        }, []);
+    const isFilledButton = ButtonType.FILLED === (buttonType as string);
```
Comment (katerynakostikova) [inline mobile/src/bundles/common/components/button/button.tsx]:
```suggestion
    const isFilledButton = ButtonType.FILLED === buttonType
```
I don't get any linter errors

**Review by katerynakostikova — APPROVED** [review]

File: mobile/src/bundles/common/components/button/button.tsx (hunk 25 above)
Comment (mskmee) [reply mobile/src/bundles/common/components/button/button.tsx]: Strange thing, I will recheck configuration tomorrow, refactor code like it was before

**Review by katerynakostikova — APPROVED** [review]

**Review by nikita-remeslov — COMMENTED** [review]
Conflicts

File: .gitignore [hunk 26]
```diff
@@ -7,4 +7,3 @@ build
 # misc
 .DS_Store
 .env
-.idea
```
Comment (nikita-remeslov) [inline .gitignore]: you should revert this changes

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]

**Review by nikita-remeslov — APPROVED** [review]
